### PR TITLE
rosenpass: init at 0.2.0

### PR DESCRIPTION
**Pull request by wucke13**
###### Description of changes
Introduce a new package for post-quantum secure WireGuard VPN

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wucke13** [comment]
@GrahamcOfBorg build rosenpass

**Comment by wucke13** [comment]
@GrahamcOfBorg build rosenpass

**Comment by wucke13** [comment]
@Lassulus Would you mind taking a look?

File: pkgs/tools/networking/rosenpass/default.nix [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    bash # for patchShebangs to find it
+    libsodium
+  ];
+
+  # otherwise pkg-config tries to link non-existent dynamic libs during the
+  # build of liboqs
+  PKG_CONFIG_ALL_STATIC = true;
+
+  # nix defaults to building for aarch64 _without_ the armv8-a
+  # crypto extensions, but liboqs depens on these
```
Comment (dadada) [inline pkgs/tools/networking/rosenpass/default.nix:57]:
```suggestion
  # crypto extensions, but liboqs depends on these
```

File: pkgs/tools/networking/rosenpass/default.nix [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    lockFile = src + "/Cargo.lock";
+  };
+
+  nativeBuildInputs = [
+    cmake # for oqs build in the oqs-sys crate
+    makeWrapper # for the rp shellscript
+    pkg-config # let libsodium-sys-stable find libsodium
+    removeReferencesTo
+    rustPlatform.bindgenHook # for C-bindings in the crypto libs
+  ];
```
Comment (dadada) [inline pkgs/tools/networking/rosenpass/default.nix:46]:
```suggestion
  ];

```

File: pkgs/tools/networking/rosenpass/default.nix [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+  # nix defaults to building for aarch64 _without_ the armv8-a
+  # crypto extensions, but liboqs depens on these
+  preBuild = lib.optionalString targetPlatform.isAarch
+    ''NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -march=armv8-a+crypto"'';
+
+  preInstall = ''
+    install -D rp $out/bin/rp
+    wrapProgram $out/bin/rp --prefix PATH : "${ rpBinPath }"
+    ls -lah
```
Comment (dadada) [inline pkgs/tools/networking/rosenpass/default.nix:64]:
```suggestion
```

**Comment by wucke13** [comment]
@dadada thank you for the review :) @Emantor Any additions from you?

**Comment by wucke13** [comment]
@GrahamcOfBorg build rosenpass

**Comment by wucke13** [comment]
@GrahamcOfBorg build rosenpass

**Comment by dadada** [comment]
I believe the ofborg-eval failure is because of cargoLock.lockFile referencing something inside `src`. Should not be an issue if we can use `cargoHash` instead. So this could be related to #218424

**Comment by dadada** [comment]
liboqs did a release with our changes. This means we only have to update Rosenpass to liboqs >=0.8

File: pkgs/tools/networking/rosenpass/default.nix [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  rpBinPath = lib.makeBinPath [
+    coreutils
+    findutils
+    gawk
+    wireguard-tools
+  ];
+in
+rustPlatform.buildRustPackage rec {
+  pname = "rosenpass";
+  version = "0.1.1";
```
Comment (dadada) [inline pkgs/tools/networking/rosenpass/default.nix:27]:
```suggestion
  version = "0.2.0";
```

File: pkgs/tools/networking/rosenpass/default.nix [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  src = fetchFromGitHub {
+    owner = pname;
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-KU/FfiaDqtAxg6/66PrI1zJ3SZBnnVUYOlc24WYJNLY=";
```
Comment (dadada) [inline pkgs/tools/networking/rosenpass/default.nix:32]:
```suggestion
    sha256 = "sha256-r7/3C5DzXP+9w4rp9XwbP+/NK1axIP6s3Iiio1xRMbk=";
```

File: pkgs/tools/networking/rosenpass/default.nix [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  };
+
+  cargoLock.lockFile = src + "/Cargo.lock";
```
Comment (dadada) [inline pkgs/tools/networking/rosenpass/default.nix:35]:
```suggestion
  cargoHash = "sha256-g2w3lZXQ3Kg3ydKdFs8P2lOPfIkfTbAF0MhxsJoX/E4=";
```

**Comment by dadada** [comment]
With version 0.2.0 at least one of the tests is failing with SIGSEGV.

```
Caused by:
  process didn't exit successfully: `/build/source/target/x86_64-unknown-linux-gnu/release/deps/rosenpass-68cd96ed62689f5a --test-threads=16` (signal: 11, SIGSEGV: invalid memory reference)
error: builder for '/nix/store/imx4vng4905qc8pmzb9p2phg0q32mnkk-rosenpass-0.2.0.drv' failed with exit code 101;
       last 10 log lines:
       > test coloring::test::sodium_malloc ... ok
       > test coloring::test::secret_memory_pool_drop ... ok
       > test config::test::test_simple_cli_parse ... ok
       > test msgs::test_constants::biscuit_ct_len ... ok
       > test msgs::test_constants::biscuit_pt_len ... ok
       > test msgs::test_constants::sodium_keysize ... ok
       > error: test failed, to rerun pass `--lib`
       >
       > Caused by:
       >   process didn't exit successfully: `/build/source/target/x86_64-unknown-linux-gnu/release/deps/rosenpass-68cd96ed62689f5a --test-threads=16` (signal: 11, SIGSEGV: invalid memory reference)
       For full logs, run 'nix log /nix/store/imx4vng4905qc8pmzb9p2phg0q32mnkk-rosenpass-0.2.0.drv'.

```

**Comment by dadada** [comment]
@wucke13 could you allow edits by reviewers, so I can bring the branch up-to-date?

**Comment by wucke13** [comment]
@dadada I think this was already the case: 
image
Anyhow, thank you for the remarks, I will quickly address them as it seems that you cannot amend to this PR

**Comment by wucke13** [comment]
@GrahamcOfBorg build rosenpass

**Comment by wucke13** [comment]
@GrahamcOfBorg build rosenpass

**Review by lorenzleutgeb — APPROVED** [review]

**Review by andres-nav — APPROVED** [review]

**Comment by Lassulus** [comment]
Result of `nixpkgs-review pr 218422` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>rosenpass</li>
  </ul>
</details>

**Pull request merged**